velox/exec/WindowPartition.cpp

```diff
@@ -130,42 +176,76 @@ bool WindowPartition::compareRowsWithSortKeys(const char* lhs, const char* rhs)
   return false;
 }
 
+vector_size_t WindowPartition::findPeerRowEndIndex(
+    vector_size_t currentStart,
+    vector_size_t lastPartitionRow,
+    const std::function<bool(const char*, const char*)> peerCompare) {
+  auto peerEnd = currentStart;
+  while (peerEnd <= lastPartitionRow) {
+    if (peerCompare(
+            partition_[currentStart - startRowOffset()],
+            partition_[peerEnd - startRowOffset()])) {
+      break;
+    }
+    ++peerEnd;
+  }
+  return peerEnd;
+}
+
 std::pair<vector_size_t, vector_size_t> WindowPartition::computePeerBuffers(
     vector_size_t start,
     vector_size_t end,
     vector_size_t prevPeerStart,
     vector_size_t prevPeerEnd,
     vector_size_t* rawPeerStarts,
-    vector_size_t* rawPeerEnds) const {
+    vector_size_t* rawPeerEnds) {
   const auto peerCompare = [&](const char* lhs, const char* rhs) -> bool {
     return compareRowsWithSortKeys(lhs, rhs);
   };
 
-  VELOX_CHECK_LE(end, numRows());
+  VELOX_CHECK_LE(end, numRows() + startRowOffset());
 
-  const auto lastPartitionRow = numRows() - 1;
+  auto lastPartitionRow = numRows() + startRowOffset() - 1;
   auto peerStart = prevPeerStart;
   auto peerEnd = prevPeerEnd;
-  for (auto i = start, j = 0; i < end; ++i, ++j) {
-    // When traversing input partition rows, the peers are the rows with the
-    // same values for the ORDER BY clause. These rows are equal in some ways
-    // and affect the results of ranking functions. This logic exploits the fact
-    // that all rows between the peerStart and peerEnd have the same values for
-    // rawPeerStarts and rawPeerEnds. So we can compute them just once and reuse
-    // across the rows in that peer interval. Note: peerStart and peerEnd can be
-    // maintained across getOutput calls. Hence, they are returned to the
-    // caller.
+
+  auto nextStart = start;
+
+  if (partial_ && start > 0) {
+    lastPartitionRow = end - 1;
+    const auto peerGroup = peerCompare(partition_[0], partition_[1]);
+
+    // The first row is the last row in previous batch. So Delete it after
+    // compare.
+    eraseRows(1);
+
+    if (!peerGroup) {
+      peerEnd = findPeerRowEndIndex(start, lastPartitionRow, peerCompare);
+
+      for (auto i = 0; i < (peerEnd - start); i++) {
```

Comment:
Maybe we could have peerIndex to access rawPeerStarts to simplify the implementation?